commit 039bfc457e436dc5c6b1d6ef996048ae19ff480d
Author: Andres Freund <andres@anarazel.de>
Date:   Fri Apr 25 12:31:42 2025 -0400

    aio: Improve debug logging around waiting for IOs
    
    Trying to investigate a bug report by Alexander Lakhin made it apparent that
    the debug logging around waiting for IO completion is insufficient. Fix that.
    
    Discussion: https://postgr.es/m/h4in2db37vepagmi2oz5vvqymjasc5gyb4lpqkunj4eusu274i@37jpd3c2spd3

diff --git a/src/backend/storage/aio/aio.c b/src/backend/storage/aio/aio.c
index e2a133cfac6..57b9cf3dcab 100644
--- a/src/backend/storage/aio/aio.c
+++ b/src/backend/storage/aio/aio.c
@@ -709,12 +709,14 @@ pgaio_io_reclaim(PgAioHandle *ioh)
 static void
 pgaio_io_wait_for_free(void)
 {
 	int			reclaimed = 0;
 
-	pgaio_debug(DEBUG2, "waiting for self with %d pending",
-				pgaio_my_backend->num_staged_ios);
+	pgaio_debug(DEBUG2, "waiting for free IO with %d pending, %d in-flight, %d idle IOs",
+				pgaio_my_backend->num_staged_ios,
+				dclist_count(&pgaio_my_backend->in_flight_ios),
+				dclist_is_empty(&pgaio_my_backend->idle_ios));
 
 	/*
 	 * First check if any of our IOs actually have completed - when using
 	 * worker, that'll often be the case. We could do so as part of the loop
 	 * below, but that'd potentially lead us to wait for some IO submitted
@@ -741,11 +743,16 @@ pgaio_io_wait_for_free(void)
 	 */
 	if (pgaio_my_backend->num_staged_ios > 0)
 		pgaio_submit_staged();
 
 	if (dclist_count(&pgaio_my_backend->in_flight_ios) == 0)
-		elog(ERROR, "no free IOs despite no in-flight IOs");
+		ereport(ERROR,
+				errmsg_internal("no free IOs despite no in-flight IOs"),
+				errdetail_internal("%d pending, %d in-flight, %d idle IOs",
+								   pgaio_my_backend->num_staged_ios,
+								   dclist_count(&pgaio_my_backend->in_flight_ios),
+								   dclist_is_empty(&pgaio_my_backend->idle_ios)));
 
 	/*
 	 * Wait for the oldest in-flight IO to complete.
 	 *
 	 * XXX: Reusing the general IO wait is suboptimal, we don't need to wait
@@ -1134,11 +1141,17 @@ pgaio_closing_fd(int fd)
 
 	/*
 	 * For now just submit all staged IOs - we could be more selective, but
 	 * it's probably not worth it.
 	 */
-	pgaio_submit_staged();
+	if (pgaio_my_backend->num_staged_ios > 0)
+	{
+		pgaio_debug(DEBUG2,
+					"submitting %d IOs before FD %d gets closed",
+					pgaio_my_backend->num_staged_ios, fd);
+		pgaio_submit_staged();
+	}
 
 	/*
 	 * If requested by the IO method, wait for all IOs that use the
 	 * to-be-closed FD.
 	 */
@@ -1166,10 +1179,14 @@ pgaio_closing_fd(int fd)
 			}
 
 			if (!ioh)
 				break;
 
+			pgaio_debug_io(DEBUG2, ioh,
+						   "waiting for IO before FD %d gets closed, %d in-flight IOs",
+						   fd, dclist_count(&pgaio_my_backend->in_flight_ios));
+
 			/* see comment in pgaio_io_wait_for_free() about raciness */
 			pgaio_io_wait(ioh, ioh->generation);
 		}
 	}
 }
@@ -1197,10 +1214,14 @@ pgaio_shutdown(int code, Datum arg)
 	 */
 	while (!dclist_is_empty(&pgaio_my_backend->in_flight_ios))
 	{
 		PgAioHandle *ioh = dclist_head_element(PgAioHandle, node, &pgaio_my_backend->in_flight_ios);
 
+		pgaio_debug_io(DEBUG2, ioh,
+					   "waiting for IO to complete during shutdown, %d in-flight IOs",
+					   dclist_count(&pgaio_my_backend->in_flight_ios));
+
 		/* see comment in pgaio_io_wait_for_free() about raciness */
 		pgaio_io_wait(ioh, ioh->generation);
 	}
 
 	pgaio_my_backend = NULL;
